### DISCUSSION TITLE
Error when launching EDB

### DISCUSSION THREAD

**Discussion by hui-zhou-a** [ansys/pyedb#1153] [General]
Hi @ring630, 

I'm looking deeper into PyAEDT, and I am getting an error where if I just start an EDB class, it says I am trying to access memory that could be corrupted or is protected and will crash the program. I have attached the image for your reference. I correctly installed pyaedt and have a functional ANSYS tool for SIwave and Electronics Desktop. Is there a reason why I am getting this error, and/or did I forget to potentially install something or set a path?

![EDBIssue](https://github.com/ansys/pyaedt/assets/76533708/9bcc59d7-8c31-4156-9347-39c76095116f)

_Originally posted by @loganchoi in https://github.com/ansys/pyaedt/discussions/4156#discussioncomment-8350908_

**Comment by hui-zhou-a** [comment]
@Samuelopez-ansys @svandenb-dev

**Comment by hui-zhou-a** [comment]
@loganchoi, i opened a new issue to track your second question.

**Comment by svandenb-dev** [comment]
@ring630,
I just made a test in PyEDB and all is working fine for me. I would suggest waiting we release PyEDB and hopefully it should be resolved automaticallly.
edb = Edb()
PyEDB INFO: Edb version 2024.1
PyEDB INFO: StdOut is enabled
PyEDB INFO: Logger is initialized in EDB.
PyEDB INFO: legacy v0.1.dev0
PyEDB INFO: Python version 3.9.4 (tags/v3.9.4:1f2e308, Apr  6 2021, 13:40:21) [MSC v.1928 64 bit (AMD64)]
PyEDB INFO: No EDB is provided. Creating a new EDB C:\Users\svandenb\Documents\layout_BJ98OP.aedb.
PyEDB INFO: EDB C:\Users\svandenb\Documents\layout_BJ98OP.aedb created correctly.
PyEDB INFO: EDB initialized.

**Comment by hui-zhou-a** [comment]
@loganchoi, I cannot reproduce your problem. Does your machine have access to licenser server?

**Comment by loganchoi** [comment]
@ring630 The ANSYS SIWave versions I utilize are 2022 R1 and 2023 R1 and I have already added the licenses to all the ANSYS products to my environment variables.

**Comment by hui-zhou-a** [comment]
@loganchoi by the error log, it looks like a .NET framwork related issue. One possible reason is your Python environment is corrupted. Can you create a new virtual environment and install the latest PyAEDT? and give it a try.

**Comment by Samuelopez-ansys** [comment]
Maybe the default directory does not have write permissions?

**Comment by SMoraisAnsys** [comment]
It could be related to the compatibility between the DLL and the python environment.
Btw, it seems that the test is performed in python 3.12 and we only support:

-    "Programming Language :: Python :: 3.7",
-    "Programming Language :: Python :: 3.8",
-    "Programming Language :: Python :: 3.9",
-    "Programming Language :: Python :: 3.10"

We might update this list if we ensure that the tests are working fine.

**Comment by loganchoi** [comment]
![ListError](https://github.com/ansys/pyaedt/assets/76533708/0d80238e-13c5-4c9c-88f0-1356fc796c8a)

@Samuelopez-ansys I specified a different directory where I have permissions and it doesn't throw the corrupted memory error anymore. However, it says I failed to initialize DLLs and that I have a list index out of range error. 

@SMoraisAnsys I will try that next and let you know what happened.

@ring630 I have the correct pyaedt version of 7.4 and reinstalled it as well.

**Comment by loganchoi** [comment]
@SMoraisAnsys I changed my python version to be 3.10 and I still get the same errors.

**Comment by svandenb-dev** [comment]
@loganchoi It fails to import the EDB dll. Seems you are on Windows ? 
just in case are you sure you got license available from the user terminal ?

**Comment by loganchoi** [comment]
@svandenb-dev I have the ANSYS tool licenses in my path in the environment variables. Do I need to add something else for me to access it via user terminal?

**Comment by svandenb-dev** [comment]
@loganchoi No I think it should be ok. Can you confirm you can start Desktop like HFSS from PyAEDT ? If yes you may have some conflicts in your libraries which prevent loading the EDB dll. It's not easy to figure out without more information.

**Comment by Samuelopez-ansys** [comment]
Hi @loganchoi ,

Please try:

1) Delete the virtual environment

2) Create a new virtual environment

3) Install pyaedt (pip install pyaedt)

4) Check if AEDT 2023R1 or 2023R2 is opening correctly on your computer, and check that you do not have any license issue.

5) Try the script

If after all these steps, it is still not working, please send an email: pyansys.support@ansys.com

**Comment by loganchoi** [comment]
@svandenb-dev Yeah, I can start a desktop application and start an HFSS project by using PyAEDT. 

@Samuelopez-ansys I created a new virtual environment and AEDT 2023R1 can open correctly, but can't run an Edb() command.

**Comment by loganchoi** [comment]
@Samuelopez-ansys I sent an email to pyansys.support@ansys.com and I got an automatic email saying that the email will not be monitored anymore and that the best support is to utilize this github or on discussion forums on ansys.com.

**Comment by SMoraisAnsys** [comment]
@loganchoi Thanks for your feeback. From your previous input and answers, I have the feeling that the LOC responsible for your issue is `self._active_cell = list(self.top_circuit_cells)[0]`. Can you try the following code ? It it extracted from an example and will download an aedb file that we use for the documentation. 

```python
import pyaedt

targetfile = pyaedt.downloads.download_file('edb/ANSYS-HSD_V1.aedb')
edb = pyaedt.Edb(edbpath=targetfile)
```

If this lines of code work on your side then the reason for the failure is probably related to the AEDB file that you provide. We expect it to have some cells and are unable to get one during `open_edb()`. If the code does not work, please provide the resulting error message.

**Comment by loganchoi** [comment]
@SMoraisAnsys I've gotten a similar error of PyAEDT failing to initialize DLLs.

![EDBError](https://github.com/ansys/pyaedt/assets/76533708/0eabfe6c-da98-473a-b3bb-183b60080d28)

**Comment by SMoraisAnsys** [comment]
@loganchoi With what kind of python interpreter did you created your virtual environment ? Did you created your virtual environment (the python base was installed from scratch ? microsoft installation ? something else ?

Could you try to create a new python environment with the CPython embedded with AEDT ? The path should be something like `"C:\Program Files\AnsysEM\v231\Win64\commonfiles\CPython\3_7\winx64\Release\python\python.exe"` or `"C:\Program Files\AnsysEM\v232\Win64\commonfiles\CPython\3_10\winx64\Release\python\python.exe"`.

From PowerShell, the full command to create the virtual environment would be ` &"C:\Program Files\AnsysEM\v231\Win64\commonfiles\CPython\3_7\winx64\Release\python\python.exe" -m venv VirtualEnv`. Then proceed as before.

**Comment by loganchoi** [comment]
@Samuelopez-ansys I utilized the python based to create the virtual environment. with the command 

python3.10 -m venv environment

I followed your instructions and still get the DLLs error.

**Comment by Samuelopez-ansys** [comment]
@loganchoi Could you check your virtual environments? If you have anyone related to Ansys, or pointing to any other AEDT version, please remove it.

In addition, can you install any other AEDT version (2023R2 for instance)? Maybe your AEDT version is corrupted because it is my first time seeing this issue.

**Comment by SMoraisAnsys** [comment]
> @Samuelopez-ansys I utilized the python based to create the virtual environment. with the command
> 
> python3.10 -m venv environment
> 
> I followed your instructions and still get the DLLs error.

Am I right to assume that this message was addressed to me ? If you still have the same error while using the python interpreter included in AEDT common files, I would suggestion to follow Samuel's comment. There might be an issue with your AEDT installation.

**Comment by shash-sharma** [comment]
Hi, not sure what the status of this issue is, but I'm seeing a related error to the one described by the OP when attempting to use EDB via pyAEDT or pyAEDB on Windows 11. I saw this error (copied below) when using 2022.R2, and followed all the suggestions above, including creating a fresh python environment, reinstalling Ansys. I also tried uninstalling 2022.R2 and upgrading to 2023.R2, but I see the same error.

To reproduce:

```python
import pyaedt
edb = pyaedt.Edb() # I've ensured I have all permissions for the default directory, and I've tried specifying other directories
```

Python version: 3.11.8

Error:

```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at Ansys.Ansoft.Edb.Definition.MaterialDef.EDBMaterialDef_Create(IntPtr, System.String)
--------------------------------
   at Ansys.Ansoft.Edb.Definition.MaterialDef.Create(Ansys.Ansoft.Edb.Database, System.String)
   at System.RuntimeMethodHandle.InvokeMethod(System.Object, System.Span`1<System.Object> ByRef, System.Signature, Boolean, Boolean)
   at System.Reflection.RuntimeMethodInfo.Invoke(System.Object, System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo)
   at Python.Runtime.MethodBinder.Invoke(Python.Runtime.BorrowedReference, Python.Runtime.BorrowedReference, Python.Runtime.BorrowedReference, System.Reflection.MethodBase, System.Reflection.MethodBase[])
   at Python.Runtime.MethodObject.Invoke(Python.Runtime.BorrowedReference, Python.Runtime.BorrowedReference, Python.Runtime.BorrowedReference, System.Reflection.MethodBase)
   at Python.Runtime.MethodBinding.tp_call(Python.Runtime.BorrowedReference, Python.Runtime.BorrowedReference, Python.Runtime.BorrowedReference)
```

The same thing happens when I try doing this through `pyaedb`.

I also tried @SMoraisAnsys 's suggestion of running the following code:

```python
import pyaedt

targetfile = pyaedt.downloads.download_file('edb/ANSYS-HSD_V1.aedb')
edb = pyaedt.Edb(edbpath=targetfile)
```
which gives the same error as reported by @loganchoi :

```
PyAEDT ERROR:   File "C:\Users\myname\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\pyedb\dotnet\edb.py", line 224, in __init__
PyEDB ERROR:   File "C:\Users\myname\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\pyedb\dotnet\edb.py", line 224, in __init__
PyAEDT ERROR:     self.open_edb()
PyEDB ERROR:     self.open_edb()
PyAEDT ERROR:   File "C:\Users\myname\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\pyedb\dotnet\edb.py", line 494, in open_edb
PyEDB ERROR:   File "C:\Users\myname\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\pyedb\dotnet\edb.py", line 494, in open_edb
PyAEDT ERROR: Check Online documentation on: https://edb.docs.pyansys.com/version/stable/search.html?q=open_edb
PyEDB ERROR: Check Online documentation on: https://edb.docs.pyansys.com/version/stable/search.html?q=open_edb
PyAEDT INFO: Failed to initialize DLLs.
PyEDB INFO: Failed to initialize DLLs.
```

Any suggestions would be appreciated. Thanks!

**Comment by svandenb-dev** [comment]
Hi @shash-sharma,
Would you have a chance trying to install the newest Ansys 2024.1 release ? 
We are not 100% sure but it seems a bug was fixed and we see this issue disappeared with updating ansys release.

**Comment by shash-sharma** [comment]
I can't upgrade to 2024 yet due to compatibility concerns, but I found that using Python 3.10 rather than 3.11 seems to fix this. Alternatively, using the python executable inside the Ansys folder per @SMoraisAnsys 's [comment](https://github.com/ansys/pyedb/issues/204#issuecomment-1960992280) above fixes this - which is how I realized that the python version might be the issue. Thanks!
